pkg/server/commands/reverseexpand/reverse_expand_weighted.go

```diff
@@ -0,0 +1,496 @@
+// It takes a `queryJob`, which encapsulates the current state of the traversal (found object,
+// and the reverse expand request with its relation stack).
+// The method constructs a database query based on the current relation at the top of the stack
+// and the `foundObject` from the previous step. It queries the datastore, and for each result:
+//   - If the relation stack is empty, it means a candidate object has been found, which is then sent to `resultChan`.
+//   - If matching tuples are found, it prepares new `queryJob` instances to continue the traversal further up the graph,
+//     using the newly found object as the `foundObject` for the next step.
+//   - If no matching objects are found in the datastore, this branch of reverse expand is a dead end, and no more jobs are needed.
+func (c *ReverseExpandQuery) executeQueryJob(
+	ctx context.Context,
+	job queryJob,
+	resultChan chan<- *ReverseExpandResult,
+	needsCheck bool,
+	jobDedupeMap *sync.Map,
+) ([]queryJob, error) {
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+
+	// Ensure we're always working with a copy
+	currentReq := job.req.clone()
+
+	userFilter, err := buildUserFilter(currentReq, job.foundObject)
+	if err != nil {
+		return nil, err
+	}
+
+	// Now pop the top relation off of the stack for querying
+	val, ok := currentReq.relationStack.Pop()
+	if !ok {
+		// should never happen, if there's no stack we shouldn't be in here
+		return nil, fmt.Errorf("unexpected empty stack in queryFunc")
+	}
+
+	entry, ok := val.(typeRelEntry)
+	if !ok {
+		// should never happen, we're only pushing typeRelEntry to this stack
+		return nil, fmt.Errorf("unexpected type encountered in query stack %T", val)
+	}
+	typeRel := entry.typeRel
+
+	// Ensure that we haven't already run this query
+	ok = checkQueryIsUnique(jobDedupeMap, userFilter, typeRel)
+	if !ok {
+		return nil, nil
+	}
+
+	objectType, relation := tuple.SplitObjectRelation(typeRel)
+
+	filteredIter, err := c.buildFilteredIterator(ctx, currentReq, objectType, relation, userFilter)
+	if err != nil {
+		return nil, err
+	}
+	defer filteredIter.Stop()
+
+	var nextJobs []queryJob
+
+	for {
+		tupleKey, err := filteredIter.Next(ctx)
+		if err != nil {
+			if errors.Is(err, storage.ErrIteratorDone) {
+				break
+			}
+			return nil, err
+		}
+
+		// This will be a "type:id" e.g. "document:roadmap"
+		foundObject := tupleKey.GetObject()
+
+		// If there are no more type#rel to look for in the stack that means we have hit the base case
+		// and this object is a candidate for return to the user.
+		if currentReq.relationStack.Empty() {
+			c.trySendCandidate(ctx, needsCheck, foundObject, resultChan)
+			continue
+		}
+
+		// For non-recursive relations (majority of cases), if there are more items on the stack, we continue
+		// the evaluation one level higher up the tree with the `foundObject`.
+		nextJobs = append(nextJobs, queryJob{foundObject: foundObject, req: currentReq})
```

Comment:
We treat this as a stream rather than collecting/reducing. Perhaps a subsequent PR it can tackle that.